localstack/services/s3/provider.py

```diff
@@ -1,12 +1,252 @@
+import logging
 import os
+from urllib.parse import quote
 
-from localstack.aws.api.s3 import S3Api
+import moto.s3.models as moto_s3_models
+import moto.s3.responses as moto_s3_responses
+from moto.s3 import s3_backends as moto_s3_backends
+
+from localstack.aws.accounts import get_aws_account_id
+from localstack.aws.api import CommonServiceException, RequestContext, ServiceResponse, handler
+from localstack.aws.api.s3 import (
+    AccountId,
+    BucketName,
+    ChecksumAlgorithm,
+    ContentMD5,
+    CreateBucketOutput,
+    CreateBucketRequest,
+    GetBucketLifecycleConfigurationOutput,
+    GetBucketLifecycleOutput,
+    GetObjectOutput,
+    GetObjectRequest,
+    HeadObjectOutput,
+    HeadObjectRequest,
+    InvalidBucketName,
+    LifecycleConfiguration,
+    ListObjectsOutput,
+    ListObjectsRequest,
+    ListObjectsV2Output,
+    ListObjectsV2Request,
+    NoSuchBucket,
+    NoSuchLifecycleConfiguration,
+    PutObjectOutput,
+    PutObjectRequest,
+    S3Api,
+)
+from localstack.services.moto import call_moto
 from localstack.services.plugins import ServiceLifecycleHook
+from localstack.services.s3.models import S3Store, s3_stores
+from localstack.services.s3.utils import verify_checksum
+from localstack.utils.aws import aws_stack
+from localstack.utils.objects import singleton_factory
+from localstack.utils.patch import patch
+
+LOG = logging.getLogger(__name__)
 
 os.environ[
     "MOTO_S3_CUSTOM_ENDPOINTS"
 ] = "s3.localhost.localstack.cloud:4566,s3.localhost.localstack.cloud"
 
+PATCHED_EXCEPTIONS = [NoSuchBucket, NoSuchLifecycleConfiguration]
+
+
+def get_moto_s3_backend(context: RequestContext) -> moto_s3_models.S3Backend:
+    return moto_s3_backends[context.account_id]["global"]
+
 
 class S3Provider(S3Api, ServiceLifecycleHook):
-    pass
+    @staticmethod
+    def get_store() -> S3Store:
+        return s3_stores[get_aws_account_id()][aws_stack.get_region()]
+
+    def on_after_init(self):
+        apply_moto_patches()
+
+    @handler("CreateBucket", expand=False)
+    def create_bucket(
+        self,
+        context: RequestContext,
+        request: CreateBucketRequest,
+    ) -> CreateBucketOutput:
+        bucket_name = request.get("Bucket", "")
+        validate_bucket_name(bucket=bucket_name)
+
+        response = call_moto_with_exception_patching(context, bucket_name)
+        return CreateBucketOutput(**response)
+
+    @handler("ListObjects", expand=False)
+    def list_objects(
+        self,
+        context: RequestContext,
+        request: ListObjectsRequest,
+    ) -> ListObjectsOutput:
+        bucket_name = request.get("Bucket", "")
+        response = call_moto_with_exception_patching(context, bucket_name)
```

Comment:
`call_moto` should already return a response object of the specific operation.
You could just add a type hint on the response object, manipulate it (while having type safety and hints in the IDE), and return the object again (instead of creating the response in the `return` line).